redhat-appstudio/service-provider-integration-operator
# feat: adding auth field in dockerconfigjson secret type

Signed-off-by: Pavol Baran <pbaran@redhat.com>

### What does this PR do?
$TITLE

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

fixes: https://issues.redhat.com/browse/SVPI-308

### How to test this PR?

1. Deploy with SPIO_IMG=quay.io/pbaran/spio:SVPI-308-config
2. Setup quay oauth config
3. Apply `samples/binding-dockerconfigjson.yaml` but with your repo URL which has scheme (without scheme currently doesn't work) 
4. Go through oauth flow
5. Check the Injected secret. It should be of type dockerconfigjson like this:
![image](https://user-images.githubusercontent.com/73115616/214679741-669969ff-1cda-4762-a0dd-960378819d83.png)
The data should be similar to this:
```json
{
	"auths": {
		"quay.io": {
			"auth": "JG9hdXRodG9rZW46T2RqYWdVY3E3UVpqNEhkUUo5QXJCaURxxxxxxxxcg=="
		}
	}
}
```
6. create pod based on private image from quay.io and check that it was pulled successfully:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: private-reg
spec:
  containers:
  - name: private-reg-container
    image: quay.io/<your-private-image>
  imagePullSecrets:
  - name: test-secret-dockerconfigjson
``` 

#### Running a Tekton Task:

1. Go through the steps above which should result in `test-secret-dockerconfigjson` secret.
2. Deploy Tekton to minikube:
```shell
kubectl apply --filename https://storage.googleapis.com/tekton-releases/pipeline/latest/release.yaml
```
3. Create Scopeo task. This Tekton Task copies container images from and to a remote registry.
```shell
kubectl apply -f https://api.hub.tekton.dev/v1/resource/tekton/task/skopeo-copy/0.2/raw
````
4. Create a service account for the TaskRun referencing the docker config json secret:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ServiceAccount
metadata:
  name: tekton-task-service-account
secrets:
  - name: test-secret-dockerconfigjson
EOF
```
5. Create TaskRun:
```shell
cat <<EOF | kubectl apply -f -
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  name: skopeo-run
spec:
  serviceAccountName: tekton-task-service-account
  taskRef:
    name: skopeo-copy
  params:
  - name: srcImageURL
    value: docker://quay.io/testingorganization/che-server:latest # replace with your private src img
  - name: destImageURL
    value:  docker://quay.io/testingorganization/che-server:skopeo # replace with your dest img
  workspaces:
  - name: images-url
    emptyDir: {}
EOF
```
6. The TaskRun should be successful:
![image](https://user-images.githubusercontent.com/73115616/216068597-e615eebf-dbe9-45a7-95e9-a6cfb09cf901.png)
